### ISSUE TITLE
Support local actor pools

### ISSUE THREAD

**Issue by autodidaddict** [wasmCloud/wasmCloud#54]
Today when an actor is started in a host it is only a single instance and the only way to run multiple instances is to spread them out across multiple lattice nodes. The question is whether there is any advantage to running multiple copies of the same actor in the same host. Will this provide any performance or scalability benefit?

The theory is that if we scale up the number of actor instances of the same actor within a host then we will be able to multiply that actor's throughput by that factor.

One possible way to do this would be to create a pool actor that in turn proxies calls to selected instances of the real actor. All other aspects of the system could be unaware of the difference between an actor pool and an individual actor. We will need to do some research to show that such a design actually improves performance.

**Comment by chrisUsick** [comment]
This may be a naive question but is it necessary that actors process one message at a time? I see for the http sever provider the handle_request function acquires a lock before dispatching a message. https://github.com/wasmCloud/capability-providers/blob/main/http-server/src/lib.rs
If it could handle parallel invocations then this wouldn’t be an issue.

**Comment by autodidaddict** [comment]
There's a couple of things going on here.

1. The lock that gets acquired is for the dispatcher, not for the actor on the other side of it. There is a slightly more efficient way to write this server provider without the use of that one lock, but if you look at where the lock is acquired (https://github.com/wasmCloud/capability-providers/blob/main/http-server/src/lib.rs#L212) you'll see that it's a read lock. Read locks, in Rust, are "almost" a no-op when the lock isn't held for write (the RwLock is 1 write lock `n` reads if write is not held). In other words, the use of this lock isn't actually blocking parallel dispatch.
2. Actors themselves are single-threaded by virtue of the fact that they are written in WebAssembly, which, as of the 1.1 specification, does not support the creation of or execution along threads.

So, in short - the HTTP server provider _is_ dispatching in parallel and we can illustrate this with a fairly simple load test. If the target of concurrent dispatches is the _same actor_, then those dispatches will queue up and be handled in delivery order because wasm/actors are single-threaded.

**Comment by autodidaddict** [comment]
Further, if you want a single actor to be able to handle more than one http request concurrently, then the solution is to scale up the number of running instances of that actor. 2 instances of actor A mean it can handle 2 concurrent HTTP requests.

**Comment by chrisUsick** [comment]
Thanks for the awesome response! That makes sense. Some sort of actor pool makes sense then. The main use case I image is deploying an actor and possibly having it automatically scaled so that would require being able to set the number of actors in the pool dynamically.

**Comment by autodidaddict** [comment]
Yes this feature is coming up next on our list- declarative deployments where you set the number of instances you need across a lattice of n nodes and the lattice takes care of spinning up what you need.

**Comment by autodidaddict** [comment]
This is a supported, intrinsic feature in the RC0 / OTP release.
